### ISSUE TITLE
redisson RMapCache method addAndGet throw org.redisson.client.RedisException: ERR Error running script:ERR Error running script

### ISSUE THREAD

**Issue by lovecto** [redisson/redisson#1628]
code:

```java
@Test
public void testMapCache(){
	try {
		String key = "TEST_MAP_KEY";
		RMapCache<String, Long> map = redissonClient.getMapCache("TEST_MAP");
		map.fastPutIfAbsent(key, 1L, 10, TimeUnit.SECONDS);
		if(!map.fastPutIfAbsent(key, 1L, 10, TimeUnit.SECONDS)){
			System.out.println(map.addAndGet(key, 1L));
		}	
	} catch (Exception e) {
		e.printStackTrace();
	}
}
```
throw exception like below 
```
org.redisson.client.RedisException: ERR Error running script (call to f_cb87dfaf85a65541ca30fc07170c12f219d89c54): @user_script:1: user_script:1: attempt to perform arithmetic on a nil value . channel: [id: 0xf3315638, L:/192.168.0.49:63297 - R:172.17.0.2/172.17.0.2:6390] command: (EVAL), params: [local value = redis.call('hget', KEYS[1], ARGV[2]); local expireDate = 92233720368547758; local t = ..., 8, TEST_MAP, redisson__timeout__set:{TEST_MAP}, redisson__idle__set:{TEST_MAP}, redisson_map_cache_created:{TEST_MAP}, redisson_map_cache_updated:{TEST_MAP}, redisson__map_cache__last_access__set:{TEST_MAP}, redisson_map_cache_removed:{TEST_MAP}, {TEST_MAP}:redisson_options, ...]
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:331)
	at org.redisson.client.handler.CommandDecoder.decodeCommand(CommandDecoder.java:128)
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:108)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
	at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:367)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:745)
```

but use Integer is ok 

```java
@Test
public void testMapCache(){
	try {
		String key = "TEST_MAP_KEY";
		RMapCache<String, Integer> map = redissonClient.getMapCache("TEST_MAP");
		map.fastPutIfAbsent(key, 1, 10, TimeUnit.SECONDS);
		if(!map.fastPutIfAbsent(key, 1, 10, TimeUnit.SECONDS)){
			System.out.println(map.addAndGet(key, 1));
		}	
	} catch (Exception e) {
		e.printStackTrace();
	}
}
```

**Comment by jackygurui** [comment]
I can confirm the bug can be triggered by the following code
```java
RMapCache<String, Long> map = redisson.getMapCache("testMap");
map.put("test", 1L);
map.addAndGet("test", 1L);
```

**Comment by jackygurui** [comment]
Dismissed. 

In short, this is a setup error relates to the MapEncoder.

TL;DR

By default, redisson uses Jackson to encode the value and java's Long data type is not native to JSON as oppose to Integer, thus the encoded value is not a numeric value presented in Redis, instead, it is stored as a JSON string. On the other hand, the script invoked by `addAndGet` implicitly requires the value as a Lua numeric value. Therefore the script throws an error when the condition is not met.


The correct setup to use `addAndGet` should be as follows:

```java
RMapCache<String, Long> map = redisson.getMapCache("testMap", StringCodec.INSTANCE);
map.put("test", 1L);
map.addAndGet("test", 1L);
```
